### PR TITLE
Resolve paths in list / dict keys

### PR DESCRIPTION
**Pull request by anukul**
See tests

### CODE REVIEW

File: storyruntime/utils/Dict.py [hunk 1]
```diff
@@ -13,48 +13,41 @@ def set(_dict, keys, output):
...
-            return s['int']
+            return item
+        elif isinstance(item, str):
+            # backwards-compatibility
+            return int(item)
+        elif isinstance(item, dict):
+            return Resolver.resolve(item, context)
         else:
-            raise Exception(f'Unable to parse {type(s)} as int.')
+            raise Exception(f'Unable to parse {type(item)} as int.')
```
Comment (wilzbach) [inline storyruntime/utils/Dict.py]: "as index"?

File: storyruntime/utils/Dict.py [hunk 2]
```diff
@@ -13,48 +13,41 @@ def set(_dict, keys, output):
...
-            # backwards-compatibility
-            return int(item)
-        else:
-            # general purpose dict
-            # ss output will never hit this branch
-            return item
+        if isinstance(item, str):
+            if item.lstrip('+-').isdigit():
+                # backwards-compatibility
+                return int(item)
```
Comment (wilzbach) [inline storyruntime/utils/Dict.py]: Looks like we don't need this anymore. It's not covered.

File: storyruntime/utils/Dict.py [hunk 3]
```diff
@@ -13,48 +13,41 @@ def set(_dict, keys, output):
...
+            else:
+                # general purpose dict
+                # ss output will never hit this branch
+                return item
+        elif isinstance(item, dict):
+            return Resolver.resolve(item, context)
```
Comment (wilzbach) [inline storyruntime/utils/Dict.py]: Can we change this in an `else` + assert to increase coverage + robustness.

File: storyruntime/utils/Dict.py (hunk 2 above)
Comment (anukul) [reply storyruntime/utils/Dict.py]: Yeah, because we can't cover older SS syntax in the way the tests are currently set up.
Should I remove this?

Comment (wilzbach) [reply storyruntime/utils/Dict.py]: I would say so because this has been removed a long time ago, but defer to @jude.

Comment (judepereira) [reply storyruntime/utils/Dict.py]: Feel free to remove now, it's been quite a while.

**Comment by judepereira** [comment]
Nice PR - should merge quickly.

**Comment by anukul** [comment]
@judepereira @wilzbach Can we merge this? I have some changes to make that overwrite this code.

File: storyruntime/utils/Dict.py [hunk 4]
```diff
@@ -13,48 +13,36 @@ def set(_dict, keys, output):
...
+    def parse_list_index(item, context):
+        if isinstance(item, int):
+            # generic dict
+            return item
+        elif isinstance(item, dict):
+            # storyscript $OBJECT
+            return Resolver.resolve(item, context)
         else:
-            raise Exception(f'Unable to parse {type(s)} as int.')
+            raise Exception(f'Unable to parse {type(item)} as list index.')
```
Comment (wilzbach) [inline storyruntime/utils/Dict.py]: This is still uncovered. Hence, replace the above elif with an else + `assert isinstance..`.

File: storyruntime/utils/Dict.py [hunk 5]
```diff
@@ -13,48 +13,36 @@ def set(_dict, keys, output):
...
-            return int(s)
-        elif isinstance(s, int):
-            # general purpose dict
-            # ss output will never hit this branch
-            return s
-        elif isinstance(s, dict) and s.get('$OBJECT') == 'int':
-            return s['int']
+    def parse_list_index(item, context):
+        if isinstance(item, int):
+            # generic dict
```
Comment (wilzbach) [inline storyruntime/utils/Dict.py]: Should this be plain-old index?

File: storyruntime/utils/Dict.py [hunk 6]
```diff
@@ -13,48 +13,36 @@ def set(_dict, keys, output):
...
-            elif object_type == 'path':
-                return Resolver.path(item['paths'], context)
-        elif isinstance(item, str) and item.lstrip('+-').isdigit():
-            # backwards-compatibility
-            return int(item)
-        else:
-            # general purpose dict
-            # ss output will never hit this branch
+        if isinstance(item, str):
+            # generic dict
```
Comment (wilzbach) [inline storyruntime/utils/Dict.py]: Shouldn't this be called plain-old key?

**Comment by anukul** [comment]
Made some changes, will lift up the resolution logic outside of `Dict` in a subsequent PR.

**Comment by judepereira** [comment]
@anukul looks like it's possible to increase coverage: https://codecov.io/gh/storyscript/runtime/pull/442/diff

**Comment by judepereira** [comment]
Looks like Codecov isn't displaying the right info. Running local coverage in IntelliJ to see actuals.

**Pull request merged**